### ISSUE TITLE
Kaifa e0025 (115200 7E1) not supported?

### ISSUE THREAD

**Issue by Gr33nG081** [openhab/openhab#4175]
Hi,

I don't get any readings in OpenHAB 1.8 with the DSMR binding. The smart meter I have is the Kaifa e0025. I tried to connect with Minicom to read the values. After some trial and errors I found out that the e0025 uses following settings:

> Baud rate: 115200
> Data bits: 7
> Parity: Even
> Stop bits: 1

Are these settings supported? Or does the binding needs to be updated?

I'm not sure what DSMR version it uses. First line says: _1-3:0.2.8(42)_ Sow, maybe the version is 4.2?

If you need more information, please let me know.
##

**Comment by Gr33nG081** [comment]
@mvolaart I saw your name in another similar issue. Hope you can help me out here.

Raw output

> /KFM5KAIFA-METER
> 
> 1-3:0.2.8(42)
> 0-0:1.0.0(160315195248W)
> 0-0:96.1.1(4530303235303030303336313836323135)
> 1-0:1.8.1(000234.402*kWh)
> 1-0:1.8.2(000213.314*kWh)
> 1-0:2.8.1(000000.000*kWh)
> 1-0:2.8.2(000000.000*kWh)
> 0-0:96.14.0(0002)
> 1-0:1.7.0(00.741*kW)
> 1-0:2.7.0(00.000*kW)
> 0-0:96.7.21(00009)
> 0-0:96.7.9(00005)
> 1-0:99.97.0(1)(0-0:96.7.19)(000101000001W)(2147483647*s)
> 1-0:32.32.0(00000)
> 1-0:32.36.0(00000)
> 0-0:96.13.1()
> 0-0:96.13.0()
> 1-0:31.7.0(003*A)
> 1-0:21.7.0(00.741*kW)
> 1-0:22.7.0(00.000*kW)
> 0-1:24.1.0(003)
> 0-1:96.1.0(4730303139333430323339363231313135)
> 0-1:24.2.1(160315190000W)(00426.101*m3)
> !3FB0

**Comment by mvolaart** [comment]
It is indeed very likely the DSMR version is 4.2. Unfortenately i can not find an open specification DSMR version 4.2 (i have 4.0.4 and 5.0).
I have looked into the raw data and all data is specified within the 4.04 version. So no problem here.

The reason it doesn't work is indeed the connection settings. The current version binding supports only 9600 7E1 and 115200 8N1. I will make a change in the next weeks so you can configure connection settings manually. If it is ready i will come back

**Comment by Gr33nG081** [comment]
It's indeed only the data bits that's causing the problem. I'm very grateful to hear you will update the binding soon. Thanks in advance!

**Comment by mvolaart** [comment]
I have updated the binding. Can you please test the attached development version of the binding?

This binding support an additional configuration parameter in openhab.cfg
The configuration parameter is dsmr:portsettings
The format of this setting is <speed> <databits><parity><stopbits>
<databits>: 5, 6, 7, 8
<parity>:'O' for Odd, 'E' for Even and 'N' for none
<stopbits>: 1, 2 or 1_5

Setting this parameter will disable the autodetection mechanism and always you the configured port parameters.
In your case the configuration parameter should be: 115200 7E1

[org.openhab.binding.dsmr-1.9.0-SNAPSHOT.zip](https://github.com/openhab/openhab/files/218474/org.openhab.binding.dsmr-1.9.0-SNAPSHOT.zip)

Please let me know if this solved your problem. Then i will merge the feature in GIT

Thanks in advance

**Comment by Gr33nG081** [comment]
Thanks for the update!

I'm trying to test the binding but I'm getting an error message in the log. Not sure if this is related to the binding, but maybe you do. 

`2016-04-14 16:24:18.281 [ERROR] [binding.dsmr.internal.DSMRPort] - Could not open port: /dev/ttyUSB0
gnu.io.NoSuchPortException: null
        at gnu.io.CommPortIdentifier.getPortIdentifier(CommPortIdentifier.java:273) ~[nrjavaserial-3.9.3.jar:na]
        at org.openhab.binding.dsmr.internal.DSMRPort.open(DSMRPort.java:328) [bundlefile:na]
        at org.openhab.binding.dsmr.internal.DSMRPort.handlePortState(DSMRPort.java:259) [bundlefile:na]
        at org.openhab.binding.dsmr.internal.DSMRPort.read(DSMRPort.java:190) [bundlefile:na]
        at org.openhab.binding.dsmr.internal.DSMRBinding.execute(DSMRBinding.java:153) [bundlefile:na]
        at org.openhab.core.binding.AbstractActiveBinding$BindingActiveService.execute(AbstractActiveBinding.java:156) [org.openhab.core_1.8.2.jar:na]
        at org.openhab.core.service.AbstractActiveService$RefreshThread.run(AbstractActiveService.java:173) [org.openhab.core_1.8.2.jar:na]
2016-04-14 16:24:18.290 [WARN ] [binding.dsmr.internal.DSMRPort] - Could not open DSMRPort, no values will be read
2016-04-14 16:24:18.291 [INFO ] [binding.dsmr.internal.DSMRPort] - Closing DSMR port` 

The port is /dev/ttyUSB0 (tested with minicom and getting the telegram readings).

Do you know if this has anything to do with the binding, or do I need to solve this issue on another level?

**Comment by mvolaart** [comment]
Are you sure ttyUSB0 exists? The serial port interface is complaining it does not exists. 

Of may be an permission issue?

**Comment by Gr33nG081** [comment]
Nailed it!

I added a line to the startup script (same as I use for my RaZberry module, only then with ttyUSB0).

Thanks for your help!

I wrote this down in my instructions:
Add line to the openhab startup script
sudo nano /etc/init.d/openhab
After the line "-Dosgi.noShutdown=true \ " add:
Dgnu.io.rxtx.SerialPorts=/dev/ttyUSB0 \

**Comment by mvolaart** [comment]
Do you also receive data with this binding?

**Comment by Gr33nG081** [comment]
I only can get one of the bindings/ports to work. So after my 'Nailed it'-moment yesterday, I found out that my Zwave wasn't working anymore. After some analyzing I'm now at the point to go search for the right fix. Hope you can get me started and else I will take a look tomorrow. 

Also I found a warning in the log about the DSMR binding. I appears in al of the options I tried. Do you know what it means and how to fix it?

ttyUSB0 in startup script (/bin/openhab.sh):
Getting readings from my smart meter.
```
2016-04-15 16:24:04.457 [WARN ] [o.config.core.ConfigDispatcher] - Cannot parse line 'dsmr:115200 7E1' of main configuration file '/etc/openhab/configurations/openhab.cfg'.
2016-04-15 16:24:31.056 [ERROR] [b.z.i.protocol.ZWaveController] - Serial Error: Port /dev/ttyAMA0 does not exist
```

ttyAMA0 in startup script (/bin/openhab.sh):
RaZberry works.
```
2016-04-15 16:28:23.846 [WARN ] [o.config.core.ConfigDispatcher] - Cannot parse line 'dsmr:115200 7E1' of main configuration file '/etc/openhab/configurations/openhab.cfg'.
2016-04-15 16:29:25.845 [ERROR] [binding.dsmr.internal.DSMRPort] - Could not open port: /dev/ttyUSB0
```


ttyAMA0 and ttyUSB0 in startup script (/bin/openhab.sh):
DSMR works, RaZberry doesnt work.
```
2016-04-15 16:32:45.298 [WARN ] [o.config.core.ConfigDispatcher] - Cannot parse line 'dsmr:115200 7E1' of main configuration file '/etc/openhab/configurations/openhab.cfg'.
2016-04-15 16:33:11.732 [ERROR] [b.z.i.protocol.ZWaveController] - Serial Error: Port /dev/ttyAMA0 does not exist
```

**Comment by mvolaart** [comment]
do you have the following parameter in openhab.cfg?
dsmr:portsettings=115200 7E1

**Comment by Gr33nG081** [comment]
I had dsmr:115200 7E1 instead of dsmr:portsettings=115200 7E1
The warn is gone now.

But I still can only use the binding if I add the line -Dgnu.io.rxtx.SerialPorts=/dev/ttyUSB0 \ to my startup script. Kind of weird, right? 

This is in the log when I don't add the line to the startup script.
```
2016-04-15 22:50:36.087 [ERROR] [binding.dsmr.internal.DSMRPort] - Could not open port: /dev/ttyUSB0
gnu.io.NoSuchPortException: null
        at gnu.io.CommPortIdentifier.getPortIdentifier(CommPortIdentifier.java:273) ~[nrjavaserial-3.9.3.jar:na]
        at org.openhab.binding.dsmr.internal.DSMRPort.open(DSMRPort.java:328) [bundlefile:na]
        at org.openhab.binding.dsmr.internal.DSMRPort.handlePortState(DSMRPort.java:259) [bundlefile:na]
        at org.openhab.binding.dsmr.internal.DSMRPort.read(DSMRPort.java:190) [bundlefile:na]
        at org.openhab.binding.dsmr.internal.DSMRBinding.execute(DSMRBinding.java:153) [bundlefile:na]
        at org.openhab.core.binding.AbstractActiveBinding$BindingActiveService.execute(AbstractActiveBinding.java:156) [org.openhab.core_1.8.2.jar:na]
        at org.openhab.core.service.AbstractActiveService$RefreshThread.run(AbstractActiveService.java:173) [org.openhab.core_1.8.2.jar:na]
2016-04-15 22:50:36.091 [WARN ] [binding.dsmr.internal.DSMRPort] - Could not open DSMRPort, no values will be read
2016-04-15 22:50:36.092 [INFO ] [binding.dsmr.internal.DSMRPort] - Closing DSMR port

```

**Comment by mvolaart** [comment]
I may have found a possible solution for your problem where i set the gnu.io.SerialPort property from the binding itself. I have added it to a test binding i attached to this message. 
Since i don't have this problem i would like to ask if you can test this for me.

Will you please make sure the log level of the dsmr binding is set to DEBUG? Can you please test it without the -Dgnu.io.txtx.SerialPorts setting and post the output in this ticket?

Thanks in advance 

[org.openhab.binding.dsmr-1.9.0-SNAPSHOT.zip](https://github.com/openhab/openhab/files/222053/org.openhab.binding.dsmr-1.9.0-SNAPSHOT.zip)

**Comment by Gr33nG081** [comment]
Okay, I think we are getting somewhere. Both bindings seem to work. Still got some issues with my zwave nodes, but since I did a clean install on my new pi (Rpi2 - upgraded from the Rpi1), I have some issues with the losing connection with the nodes. 

For addition information, I use this raspbian version: 2016-03-18-raspbian-jessie.img

Here is the OpenHAB logfile with the logback.xml settings to DEBUG.

[openhab.log.txt](https://github.com/openhab/openhab/files/222236/openhab.log.txt)

**Comment by mvolaart** [comment]
As for the DSMR binding i see indeed that it is working correctly. Thanks for the log file. I see indeed that the fix is working.

Unfortunately I can not help you with the Z-Wave items since i don't maintain this binding.

I will merge the changes so the fixes will be be part of future OpenHAB releases.

**Comment by Gr33nG081** [comment]
Thanks for updating the binding, and your help!

I understand that your can't help me out with the zwave issue. I will try to find a solution for that somewhere else.

Thanks again.

**Comment by edwinmat** [comment]
Recently Stedin installed a Kaifa smart meter and I ran into the same problems as described above. Finally I discovered the problems did not occur running openhab as root. After adding the openhab user to the dialout group and rebooting, the problem was solved for me.  I thought this is good to share...

Thanks for the good work!

**Comment by watou** [comment]
Closed via #4854
